File: cpp_draft/ft_linear_regression.cpp
//c++ -std=c++20 -I/usr/local/include -L/usr/local/lib -lmatplot ft_linear_regression.cpp -o ft_linear_regression

#ifndef LINEARMODEL_CPP
# define LINEARMODEL_CPP

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <matplot/matplot.h>
#include "normalize.cpp"

class LinearModel{
	private:
		double	_intercept;
		double	_slope;
		double	_lr; //learning rate
		std::vector<int>	_mileage;
		std::vector<double>	_n_mileage;
		std::vector<int>	_price;
		std::vector<double>	_n_price;
		std::vector<double> _n_slopes;
		std::vector<double> _n_intercepts;

		double	ssr(double slope, double intercept);

	public:
		LinearModel() : _intercept(0), _slope(0), _lr(0.01) {};
		~LinearModel() {};
		double	estimatePrice(double mileage);
		void	trainModel(const std::string &data_csv);
		void	readCsv(const std::string &data_csv);
		void	plot_data();
		void	plot_ssr();
} ;

void	LinearModel::plot_data() {
	using namespace matplot;
	
	//converting to double
    std::vector<double> mileage(_mileage.begin(), _mileage.end());
    std::vector<double> price(_price.begin(), _price.end());
    scatter(mileage, price);

	// Calculate the range for x values of the line
	auto min_max = std::minmax_element(_mileage.begin(), _mileage.end());
	double x_min = *min_max.first;
	double x_max = *min_max.second;

	// Calculate the y values for the regression line
	std::vector<double> line_x = {x_min, x_max};
	std::vector<double> line_y = {_slope * x_min + _intercept, _slope * x_max + _intercept};
    
	hold(on);

    // Plot the regression line
    plot(line_x, line_y, "b-"); // "b-" is the blue line style

	title("Training Data");
	xlabel("Mileage in km");
    ylabel("Price");
    show();
}

double	LinearModel::ssr(double slope, double intercept) {
	double	ssr = 0;
	for (int i = 0; i < _mileage.size(); i ++) {
		ssr += pow(_price[i] - (intercept + slope * _mileage[i]), 2);
	}
	return ssr;
}

void	LinearModel::plot_ssr() {
	using namespace matplot;

	std::vector<double>	slopes = linspace(-2, 2, 100);;
	std::vector<double>	intercepts = linspace(-100000, 100000, 100);;

	// Generate meshgrid
    auto [X, Y] = meshgrid(intercepts, slopes);

    // Compute Z as the sum of squared residuals for each (slope, intercept) pair
    std::vector<std::vector<double>> Z(intercepts.size(), std::vector<double>(slopes.size()));
    for (size_t i = 0; i < intercepts.size(); ++i) {
        for (size_t j = 0; j < slopes.size(); ++j) {
            Z[i][j] = ssr(slopes[j], intercepts[i]);
        }
    }

    // Create a 3D surface plot
    mesh(X, Y, Z);
    title("Surface Plot of Sum of Squared Residuals");
    ylabel("Slope");
    xlabel("Intercept");
    zlabel("Sum of Squared Residuals");

    show();
}

double	LinearModel::estimatePrice(double mileage) {
	return (this->_intercept + this->_slope * mileage);
}

void	LinearModel::readCsv(const std::string &data_csv){
	std::ifstream file(data_csv);
	std::string line;
	bool isFirstLine = true;

	while (std::getline(file, line)) {
		std::stringstream ss(line);
		std::string value;
		
		if (isFirstLine) {
			isFirstLine = false;
			continue;
		}

		std::getline(ss, value, ',');
		_mileage.emplace_back(std::stoi(value));
		std::getline(ss, value, ',');
		_price.emplace_back(std::stoi(value));		
		std::cout << _mileage.back() << " | " << _price.back() << std::endl;
	}
}

void	LinearModel::trainModel(const std::string &data_csv) {
	std::size_t	data_size;

	int			learning_iterations = 1000;
	double		residual;
	double		grad_inter, grad_slo;

	readCsv(data_csv);
	data_size = _mileage.size();
	z_score_normalize(_mileage, _price, _n_mileage, _n_price);

	for (int i = 0; i < learning_iterations; i++) {
		grad_inter = 0.0;
		grad_slo = 0.0;
		//calculate the derivation of the sum of the squared residuals with respect to intercept and slope
		//2 or more derivatives of the same functions are called gradient
		for (int j = 0; j < data_size; j++) {
			residual = _n_price[j] - estimatePrice(_n_mileage[j]);
			grad_inter += residual;
			grad_slo += residual * _n_mileage[j]; 
		}
		//update intersect and slope
		grad_inter *= -2;
		grad_slo *= -2;

		if (fabs(grad_inter) < 0.001 && fabs(grad_slo) < 0.001) {
			this->_slope = denormalize_slope(this->_slope, _mileage, _price);
			this->_intercept = denormalize_intercept(this->_intercept, this->_slope, _mileage, _price);
			return;
		}
		this->_intercept -= this->_lr * grad_inter;
		this->_slope -= this->_lr * grad_slo;
		//for plotting ssr
		_n_slopes.emplace_back(_slope);
		_n_intercepts.emplace_back(_intercept);
	}
	this->_slope = denormalize_slope(this->_slope, _mileage, _price);
	this->_intercept = denormalize_intercept(this->_intercept, this->_slope, _mileage, _price);
}

int	main() {

	LinearModel lm;
	std::cout << "estimate price for 84000 km before training: " << std::endl;
	std::cout << lm.estimatePrice(84000) << std::endl;
	lm.trainModel("data.csv");
	std::cout << "estimate price for 60949 km after training: " << std::endl;
	std::cout << lm.estimatePrice(60949) << std::endl;
	std::cout << "estimate price for 185530 km after training: " << std::endl;
	std::cout << lm.estimatePrice(185530) << std::endl;
	std::cout << "estimate price for 0 km after training: " << std::endl;
	std::cout << lm.estimatePrice(0) << std::endl;
	//lm.plot_data();
	//lm.plot_ssr();
}

#endif

File: ft_linear_regression.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

class LinearModel:
	def __init__(self, lr=0.01, learning_iterations=1000, slope=0, intercept=0):
		"""
		Initialize the LinearModel with learning rate, number of iterations, and initial slope/intercept.
		"""
		self.lr = lr
		self.learning_iterations = learning_iterations
		self.slope = slope
		self.intercept = intercept
		self.df = None
		self.df_normalized = None
		self.mean_mileage = 0
		self.std_mileage = 0
		self.mean_price = 0
		self.std_price = 0

	def estimate_price(self, mileage):
		"""
		Estimate the price based on mileage using the current model parameters.
		"""
		return self.slope * mileage + self.intercept

	def normalize_data(self):
		"""
		Normalize the data by standardizing the mileage and price columns.
		"""
		self.mean_mileage = self.df['km'].mean()
		self.std_mileage = self.df['km'].std()
		self.mean_price = self.df['price'].mean()
		self.std_price = self.df['price'].std()
		self.df_normalized = (self.df - self.df.mean()) / self.df.std()
	
	def denormalize_slope(self, slope):
		"""
		Denormalize the slope to the original scale.
		"""
		return slope * (self.std_price / self.std_mileage)

	def denormalize_intercept(self, intercept, slope):
		"""
		Denormalize the intercept to the original scale.
		"""
		return intercept * self.std_price + self.mean_price - slope * self.mean_mileage

	def calculate_gradients(self):
		"""
		Calculate the gradients for the intercept and slope based on normalized data.
		"""
		residuals = self.df_normalized['price'] - self.estimate_price(self.df_normalized['km'])
		grad_intercept = -2 * residuals.sum()
		grad_slope = -2 * (residuals * self.df_normalized['km']).sum()
		return grad_intercept, grad_slope
	
	def plot_data(self):

		# Scatter plot of original data
		plt.scatter(self.df['km'], self.df['price'], color='blue', label='Data')

		# Plotting the predicted prices (regression line)
		plt.plot(self.df['km'], self.slope * self.df['km'] + self.intercept, color='red', label='Regression Line')

		# Adding labels and title
		plt.xlabel('Mileage (km)')
		plt.ylabel('Price')
		plt.title('Linear Regression: Mileage vs Price')
		plt.legend()

		# Show plot
		plt.show()

	def ssr(self, slope, intercept):
		return np.sum((self.df['price'] - (intercept + slope * self.df['km'])) ** 2)
	
	def plot_ssr(self):

		# Define ranges for slopes and intercepts
		slopes = np.linspace(-2, 2, 100)
		intercepts = np.linspace(-100000, 100000, 100)

		# Generate meshgrid
		X, Y = np.meshgrid(intercepts, slopes)

		# Compute Z as the sum of squared residuals for each (slope, intercept) pair
		Z = np.zeros_like(X)
		for i in range(len(intercepts)):
			for j in range(len(slopes)):
				Z[j, i] = self.ssr(slopes[j], intercepts[i])

		# Create a 3D surface plot
		fig = plt.figure()
		ax = fig.add_subplot(111, projection='3d')
		ax.plot_surface(X, Y, Z, cmap='viridis')

		ax.set_title("Surface Plot of Sum of Squared Residuals")
		ax.set_xlabel("Intercept")
		ax.set_ylabel("Slope")
		ax.set_zlabel("Sum of Squared Residuals")

		plt.show()
		

	def train(self, data):
		"""
		Train the linear model using gradient descent.
		"""
		self.df = pd.read_csv(data)

		if 'km' not in self.df.columns or 'price' not in self.df.columns:
			raise ValueError("DataFrame must contain 'km' and 'price' columns")

		if self.df.empty:
			raise ValueError("DataFrame is empty")

		self.normalize_data()

		for _ in range(self.learning_iterations):
			grad_intercept, grad_slope = self.calculate_gradients()

			if abs(grad_intercept) < 0.001 and abs(grad_slope) < 0.001:
				break
			
			self.intercept -= self.lr * grad_intercept
			self.slope -= self.lr * grad_slope
		
		# Final denormalization
		self.slope = self.denormalize_slope(self.slope)
		self.intercept = self.denormalize_intercept(self.intercept, self.slope)

# Example usage
model = LinearModel(lr=0.01, learning_iterations=1000)
model.train('data.csv')

print(model.estimate_price(60949))
model.plot_ssr()